main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala

```diff
@@ -31,16 +34,137 @@ import ca.uwaterloo.flix.util.Validation._
   * (b) Appears in a fact or a rule as a filter/transfer function.
   * (c) Appears in a function which itself is reachable.
   */
+
 object TreeShaker extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
 
   /**
     * Performs tree shaking on the given AST `root`.
     */
   def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
 
-    // TODO: Implement.
+    /**
+      * A queue of function definitions to be processed recursively.
+      *
+      * For example, if the queue contains the entry:
+      *
+      * -   f
+      *
+      * it means that the function definition f should be considered to determine new reachable functions.
+      */
+    val queue: mutable.Queue[Definition.Constant] = mutable.Queue.empty
+
+    /**
+      * A map used to collect reachable functions (symbols and the corresponding definitions).
+      */
+    val reachableFunctions: mutable.Map[Symbol.DefnSym, Definition.Constant] = mutable.Map.empty
+
+    /**
+      * Returns true iff the function definition `defn` is initially reachable.
+      *
+      * That is, returns true iff `defn` satisfies one of the following:
+      *
+      *   (a) Appears in the global namespaces, takes zero arguments, and is not marked as synthetic.
+      *
+      *   (b) Appears in a fact or a rule as a filter/transfer function.
+      */
+    def firstReachable(defn: Definition.Constant): Boolean = {
```

Comment:
I'm unsure how to check for "Appears in a fact or a rule as a filter/transfer function".

The following code will break the current `TreeShaker`:
`rel R(x: Int, y: Int, z: Str)` `R(1, Int32.min(6, 2), "foo").`

This happens because the `SimplifiedAst.root` contains the entry `def autogenerated$7714() = 
  Int32.min(6i32, 2i32, )` which is marked as synthetic and thus not considered as an initially reachable function.

So should synthetic functions be considered initially reachable?  Or should any `autogenerated` function be considered initially reachable?